components/support/nimbus-fml/src/parser.rs

```diff
@@ -563,63 +588,200 @@ fn get_typeref_from_string(
+
+        // Child must not specify any features, objects or enums that the parent has.
+        let features = merge_map(
+            &parent.features,
+            &child.features,
+            "Features",
+            "features",
+            child_path,
+        )?;
+
+        let p_types = &parent.legacy_types.unwrap_or(parent.types);
+        let c_types = &child.legacy_types.unwrap_or(child.types);
+
+        let objects = merge_map(
+            &c_types.objects,
+            &p_types.objects,
+            "Objects",
+            "objects",
+            child_path,
+        )?;
+        let enums = merge_map(&c_types.enums, &p_types.enums, "Enums", "enums", child_path)?;
+
+        let merged = ManifestFrontEnd {
             features,
-            channels: manifest.channels,
-            channel,
-        })
+            types: Types { enums, objects },
+            legacy_types: None,
+            ..parent
+        };
+
+        Ok(merged)
     }
 
-    pub fn get_intermediate_representation(&self) -> Result<FeatureManifest, FMLError> {
+    pub fn get_intermediate_representation(
+        &self,
+        channel: &str,
+    ) -> Result<FeatureManifest, FMLError> {
+        let manifest = &self.manifest;
+        let enums = manifest.get_enums();
+        let objects = manifest.get_objects();
+
         let object_map: HashMap<String, &ObjectDef> =
-            self.objects.iter().map(|o| (o.name(), o)).collect();
-        let merger = DefaultsMerger::new(object_map);
+            objects.iter().map(|o| (o.name(), o)).collect();
+        let merger = DefaultsMerger::new(object_map, manifest.channels.clone(), channel.to_owned());
+
+        let features = manifest.get_feature_defs(&merger)?;
 
-        let features = self
-            .features
-            .iter()
-            .map(|feature_def| {
-                merger.merge_feature_defaults(feature_def.clone(), &self.channel, &self.channels)
-            })
-            .collect::<Result<Vec<_>, FMLError>>()?;
         Ok(FeatureManifest {
-            enum_defs: self.enums.clone(),
-            obj_defs: self.objects.clone(),
+            enum_defs: enums,
+            obj_defs: objects,
             hints: HashMap::new(),
             feature_defs: features,
         })
     }
 }
 
-#[derive(Debug, Deserialize)]
+fn check_can_merge_manifest(
+    parent: &ManifestFrontEnd,
+    child: &ManifestFrontEnd,
+    child_path: &dyn Display,
+) -> Result<()> {
+    if !child.channels.is_empty() {
+        let child = &child.channels;
+        let child = child.iter().collect::<HashSet<&String>>();
+        let parent = &parent.channels;
+        let parent = parent.iter().collect::<HashSet<&String>>();
+        if !child.is_subset(&parent) {
+            return Err(FMLError::ValidationError(
+                "channels".to_string(),
+                format!(
+                    "Included manifest should not define its own channels: {}",
```

Comment:
nit: could be a useful error message to print out which channels are the culprit, to help consumers track them down and debug